### DISCUSSION TITLE
How to trigger external self-transition to a parallel state from its sibling?

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#3517] [Q&A]
Greetings,

I expected the below code to invoke `i` twice and print `"Invoked!"` two times in the console.
How to trigger external self-transition to a parallel state (`a.x`) from its sibling (`b.x`)?
I am trying to execute the invocation again on some event (`A`).

```ts
const m = createMachine({
    id: "m",
    schema: {
        events: {} as { type: "A" }
    },
    tsTypes: {} as import("./a.test.typegen").Typegen0,

    type: "parallel",
    states: {
        a: {
            initial: "x",
            states: {
                x: {
                    invoke: {
                        id: "i",
                        src: () => () => console.log("Invoked!")
                    }
                }
            }
        },
        b: {
            initial: "y",
            states: {
                y: {
                    on: {
                        A: {
                            target: "#m.a.x",
                            internal: false
                        },
                    }
                }
            }
        }
    },
});

const s = interpret(m).start();

setTimeout(() => s.send({ type: "A" }), 500);
```

Thank you!

**Comment by davidkpiano** [comment]
I _think_ this currently doesn't work as expected because the external transition you are specifying is not for the `a.x` state, so it will not reenter that state.

What you can do is `raise(...)` an event, and handle that raised event in the `a` region to exit/reenter the `a.x` state. That is the typical way for statechart regions to communicate with each other.

**Reply by Andarist** [reply]
> I think this currently doesn't work as expected because the external transition you are specifying is not for the a.x state, so it will not reenter that state.

This transition is targeting `a.x` though - just using the ID syntax.

The issue is in our [`getExternalReentryNodes`](https://github.com/statelyai/xstate/blob/f2bf4924c291069204baab8b2f156574c0bd085f/packages/core/src/StateNode.ts#L976-L996). It only accounts for "simple" reentry nodes (when no parallel regions are involved) and assumes that if there is no parent-child relationship between source and the target states then the target will simply be entered by the logic further down the road. That logic that happens later doesn't deal with reentrancy - it just compares the previous and next configurations and computes entry/exit sets based on that. So since this state is part of the previous configuration - it's not added to the entry set.

A potential fix for this would be to fix the logic and compute the "transition domain" in the `getExternalReentryNodes` and add all states between the domain and the target to those reentry nodes. However, this is already fixed in v5 and I'm not sure if this is that important of a case to chase a fix in v4 - especially that potentially it could change behavior for some machines out there.
